File: back-end/controllers/userController.js
const axios = require('axios');

const BASE_URL = 'http://localhost:8080';

exports.registerUser = async function registerUser(req, res) {
    const { firstName, lastName, email, password } = req.body;

    try {
        const response = await axios.post(`${BASE_URL}/users`, {
            first_name: firstName,
            last_name: lastName,
            email,
            password
        });

        res.status(201).json(response.data);
    } catch (error) {
        if (error.response && error.response.status === 400) {
            res.status(400).json({ message: "Email already registered" });
        } else {
            res.status(500).json({ message: "Internal server error" });
        }
    }
};

exports.loginUser = async function loginUser(req, res) {
    const { email, password } = req.body;

    try {
        const response = await axios.post(`${BASE_URL}/sessions`, {
            email,
            password
        });

        // Set session_id cookie
        res.cookie('session_id', response.data.session_id, { httpOnly: true });
        res.status(200).json(response.data);
    } catch (error) {
        if (error.response && error.response.status === 401) {
            res.status(401).json({ message: "Invalid credentials" });
        } else {
            res.status(500).json({ message: "Internal server error" });
        }
    }
};

exports.logoutUser = async function logoutUser(req, res) {
    const session_id = req.cookies.session_id;

    try {
        const response = await axios.delete(`${BASE_URL}/sessions`, {
            headers: { Cookie: `session_id=${session_id}` }
        });

        // Clear session_id cookie
        res.clearCookie('session_id');
        res.status(200).json(response.data);
    } catch (error) {
        if (error.response && error.response.status === 403) {
            res.status(403).json({ message: "Forbidden" });
        } else {
            res.status(500).json({ message: "Internal server error" });
        }
    }
};

exports.getProfile = async function getProfile(req, res) {
    const session_id = req.cookies.session_id;

    try {
        const response = await axios.get(`${BASE_URL}/profile`, {
            headers: { Cookie: `session_id=${session_id}` }
        });

        res.status(200).json(response.data);
    } catch (error) {
        if (error.response && error.response.status === 403) {
            res.status(403).json({ message: "Forbidden" });
        } else {
            res.status(500).json({ message: "Internal server error" });
        }
    }
};

exports.requestPasswordResetToken = async function requestPasswordResetToken(req, res) {
    const { email } = req.body;

    try {
        const response = await axios.post(`${BASE_URL}/reset_password`, { email });
        res.status(200).json(response.data);
    } catch (error) {
        if (error.response && error.response.status === 403) {
            res.status(403).json({ message: "Invalid email" });
        } else {
            res.status(500).json({ message: "Internal server error" });
        }
    }
};

exports.updatePassword = async function updatePassword(req, res) {
    const { email, reset_token, new_password } = req.body;

    try {
        const response = await axios.put(`${BASE_URL}/reset_password`, {
            email,
            reset_token,
            new_password
        });

        res.status(200).json(response.data);
    } catch (error) {
        if (error.response && error.response.status === 403) {
            res.status(403).json({ message: "Invalid token or email" });
        } else {
            res.status(500).json({ message: "Internal server error" });
        }
    }
};

File: back-end/controllers/user_controller.py
#!/usr/bin/python3
"""

"""
from flask import request, Blueprint, jsonify, abort, redirect, g
from auth import Auth, login_required

user_bp = Blueprint('user_bp', __name__)
AUTH = Auth()

@user_bp.route('/users', methods=['POST'], strict_slashes=False)
def register_user() -> str:
    """
    Register user
    """
    first_name = request.form.get("first_name")
    if not first_name:
        abort(400, description="Missing first name")

    last_name = request.form.get("last_name")
    if not last_name:
        abort(400, description="Missing last name")

    email = request.form.get("email")
    if not email:
        abort(400, description="Missing email")

    password = request.form.get("password")
    if not password:
        abort(400, description="Missing password")

    try:
        AUTH.register_user(first_name, last_name, email, password)
    except ValueError:
        abort(400, description="Email already registered")
    return jsonify({"email": email, "message": "user created"})


@user_bp.route('/sessions', methods=["POST"], strict_slashes=False)
def login() -> str:
    """
    Login method
    """
    email = request.form.get("email")
    if not email:
        abort(403, description="Missing email")

    password = request.form.get("password")
    if not password:
        abort(403, description="Missing password")

    if not AUTH.valid_login(email, password):
        abort(403, desciption="Invalid Credentials")

    session_id = AUTH.create_session(email)
    response = jsonify({"email": email, "message": "logged in"})
    response.set_cookie("session_id", session_id)

    return response

@user_bp.route('/sessions', methods=["DELETE"], strict_slashes=False)
@login_required
def logout() -> str:
    """
    Logout method
    """
    session_id = request.cookies.get("session_id")
    if not session_id:
        abort(403, description="Missing session id")

    user = AUTH.get_user_from_session_id(session_id)
    if user is None:
        abort(404, description="User not found")

    AUTH.destroy_session(user.id)

    return redirect('/')

@user_bp.route('/profile', methods=['GET'], strict_slashes=False)
@login_required
def profile() -> str:
    """
    Returns the profile of the session id
    """
    user = AUTH.get_user_from_session_id(request.cookies.get('session_id'))
    if user is None:
        abort(404, description="User not found")

    return jsonify({"email": user.email, "name": f"{user.first_name} {user.last_name}"}), 200


@user_bp.route('/reset_password', methods=["POST"], strict_slashes=False)
@login_required
def get_reset_password_token() -> str:
    """
    Get reset password token
    """
    email = request.form.get('email')
    if not email:
        abort(400, description="Missing email")

    try:
        token = AUTH.get_reset_password_token(email)
        return jsonify({"email": email, "reset_token": token})
    except ValueError:
        abort(403, desciption="Invalid Credentials")


@user_bp.route('/reset_password', methods=["PUT"], strict_slashes=False)
@login_required
def update_password() -> str:
    """
    Update password
    """
    email = request.form.get('email')
    if not email:
        abort(400, description="Missing email")

    reset_token = request.form.get('reset_token')
    if not reset_token:
        abort(400, description="Missing reset token")

    new_password = request.form.get('new_password')
    if not new_password:
        abort(400, description="Missing new password")

    try:
        AUTH.update_password(reset_token, new_password)
        return jsonify({"email": email, "message": "Password updated"}), 200
    except ValueError:
        abort(403, description="Invalid credentials")

@user_bp.route('/users', methods=["DELETE"], strict_slashes=False)
@login_required
def delete_user() -> str:
    """
    Deletes a user
    """
    email = request.form.get("email")
    if not email:
        abort(400, description="Missing Email")

    password = request.form.get("password")
    if not password:
        abort(400, description="Missing password")

    try:
        AUTH.delete_user(email, password)
        return jsonify({"email": email, "message": "User deleted"})
    except ValueError as e:
        abort(403, description=e.args[0])
